R/vendor.R

```diff
@@ -0,0 +1,184 @@
+
+  if (is.null(lib_name)) {
+    lib_name <- as_valid_rust_name(pkg_name(path))
+  } else if (length(lib_name) > 1) {
+    cli::cli_abort(
+      "{.arg lib_name} must be a character scalar",
+      class = "rextendr_error"
+    )
+  }
+
+  use_rextendr_template(
+    "Makevars.vendor",
+    save_as = file.path("src", "Makevars"),
+    quiet = quiet,
+    overwrite = overwrite,
+    data = list(lib_name = lib_name)
+  )
+
+  use_rextendr_template(
+    "Makevars.win.vendor",
+    save_as = file.path("src", "Makevars.win"),
+    quiet = quiet,
+    overwrite = overwrite,
+    data = list(lib_name = lib_name)
+  )
+
+  use_rextendr_template(
+    "vendor-config.toml",
+    save_as = file.path("src", "rust", "vendor-config.toml"),
+    quiet = quiet,
+    overwrite = overwrite
+  )
+
+  # handle if usethis is not installed
+  if (!rlang::is_installed("usethis")) {
+    cli::cli_inform(
+      c(
+        "!" = "Add {.code ^src/rust/vendor$} to your {.file .Rbuildignore}",
+        "!" = "Add {.code ^src/rust/vendor$} to your {.file .gitignore}",
+        "i" = "Install {.pkg usethis} to have this done automatically."
+      )
+    )
+  } else {
+    # vendor will be big when expanded and should be ignored
+    usethis::use_build_ignore(
+      file.path("src", "rust", "vendor")
+    )
+
+    usethis::use_git_ignore(
+      file.path("src", "rust", "vendor")
+    )
+  }
+
+  invisible(NULL)
+}
+
+
+#' @export
+#' @rdname use_cargo_vendor
+vendor_pkgs <- function(path = ".",
+                        quiet = FALSE) {
+  local_quiet_cli(quiet)
+
+  # get path to rust folder
+  src_dir <- rprojroot::find_package_root_file(path, "src/rust")
+
+  # vendor crates
+  withr::with_dir(src_dir, {
+    vendor_res <- processx::run(
+      "cargo",
+      c(
+        "vendor",
+        "--locked",
+        "--manifest-path",
+        file.path(src_dir, "Cargo.toml")
+      ),
+      stderr_line_callback = function(x, proc) {
+        if (!grepl("To use vendored sources", x) && x != "") {
+          cli::cat_bullet(stringi::stri_trim_left(x))
+        }
+      }
+    )
+  })
+
+  # check to see if the config.toml is different than the template
+  # if so update it
+  config_toml <- stringi::stri_split(vendor_res$stdout, coll = "\n")[[1]]
+  # remove trailing empty string char
+  config_toml <- config_toml[1:(length(config_toml) - 1)]
+
+  # get the text of the current vendor-config.toml
+  cur_config <- readLines(file.path(src_dir, "vendor-config.toml"))
+
+  # check if they are identical
+  configs_match <- identical(
+    config_toml,
+    cur_config
+  )
+
+  # update vendor-config.toml if necessary
+  if (!configs_match) {
+    brio::write_lines(config_toml, file.path(src_dir, "vendor-config.toml"))
+    cli::cli_alert_info("Updating {.file src/rust/vendor-config.toml}")
+  }
+
+  # compress the vendor.tar.xz
+  compress_res <- withr::with_dir(src_dir, {
+    processx::run(
+      "tar", c(
+        "-cJ", "--no-xattrs", "-f", "vendor.tar.xz", "vendor"
```

Comment:
See: https://github.com/extendr/rextendr/pull/313#issuecomment-1700908011
Current behavior changes the hash (which the test is based on).
Use args from #311 to ensure higher compression and consistent hashing.

Reply:
I have confirmed hashing behavior.

Reply:
I realized that the R-core team has already made a utility to make this easier 🤦🏼 
`tar("../vendor.tar.xz", compression = "xz", compression_level = 9)`

Reply:
I think I even tried the `tar` function at the time, but for some reason gave up using it.

Perhaps rextendr should test that the tarball hash is immutable.

Reply:
I agree. Its in this test
https://github.com/extendr/rextendr/blob/8ac7cf00ccb5e8ffea1e653ea8162170c6b18d10/tests/testthat/test-use_cargo_vendor.R#L30

Reply:
The `tar()` function is very weird....it expands in a very weird way. I've reverted so it will compile—i know it will work on ubuntu, windows, and mac. This can be improved in a follow up PR.